File: icpc/《算法设计编程实验（第2版）试题原版和测试数据/第5章  贪心法的编程实验/5.4相关题库/5.4.6 Annoying painting tool/annoying.java
// Author: Adrian Kuegel

import java.io.*;
import java.util.*;

public class annoying {
	public static void main(String [] args) throws Exception {
		BufferedReader in = new BufferedReader(new FileReader("annoying.in"));
		String line;
		while((line = in.readLine()) != null) {
			StringTokenizer strtok = new StringTokenizer(line);
			int n = Integer.parseInt(strtok.nextToken());
			int m = Integer.parseInt(strtok.nextToken());
			int r = Integer.parseInt(strtok.nextToken());
			int c = Integer.parseInt(strtok.nextToken());
			if (n == 0 && m == 0 && r == 0 && c == 0)
				break;
			char [][] p = new char[n][m];
			for (int i=0; i<n; ++i) {
				line = in.readLine();
				p[i] = line.toCharArray();
				for (int j=0; j<m; ++j)
					p[i][j] -= '0';
			}
			int cnt = 0;
			for (int i=0; i+r<=n; ++i)
				for (int j=0; j+c<=m; ++j) {
					if (p[i][j] == 0)
						continue;
					++cnt;
					for (int ii=0; ii<r; ++ii)
						for (int jj=0; jj<c; ++jj)
							p[i+ii][j+jj] ^= 1;
				}
			for (int i=0; i<n; ++i)
				for (int j=0; j<m; ++j)
					if (p[i][j] != 0)
						cnt = -1;
			System.out.println(cnt);
		}
	}
}

File: icpc/《算法设计编程实验（第2版）试题原版和测试数据/第5章  贪心法的编程实验/5.4相关题库/5.4.6 Annoying painting tool/annoying.cpp
// Author: Adrian Kuegel
// Date: 15. 6. 2007
// Algorithm: greedy
// Complexity: O(r*(n-r)*c*(m-c))

#include <stdio.h>
#include <assert.h>

char p[128][131];

int main() {
	FILE *in = fopen("annoying.in","r");
	int n, m, r, c;
	while(fscanf(in, "%d %d %d %d ",&n,&m,&r,&c) == 4) {
		if (n == 0 && m == 0 && r == 0 && c == 0)
			break;
		assert(1 <= r && r <= n && n <= 100 && 1 <= c && c <= m && m <= 100);
		for (int i=0; i<n; ++i) {
			fgets(p[i],sizeof(p[i]),in);
			for (int j=0; j<m; ++j) {
				assert(p[i][j] >= '0' && p[i][j] <= '1');
				p[i][j] -= '0';
			}
			assert(p[i][m] == '\n');
		}
		int cnt = 0;
		for (int i=0; i+r<=n; ++i)
			for (int j=0; j+c<=m; ++j) {
				if (!p[i][j])
					continue;
				++cnt;
				for (int ii=0; ii<r; ++ii)
					for (int jj=0; jj<c; ++jj)
						p[i+ii][j+jj] ^= 1;
			}
		for (int i=0; i<n; ++i)
			for (int j=0; j<m; ++j)
				if (p[i][j])
					cnt = -1;
		printf("%d\n",cnt);
	}
	assert(n == 0 && m == 0 && r == 0 && c == 0);
	fclose(in);
	return 0;
}
